### PR TITLE
rest-api: make request timeout configurable

### LINKED ISSUE #1833: Make xDS request timeout configurable
Async Client uses NullRetryPolicy and the API fetcher hardcodes the upstream request timeout to 1s. This is too low in some environments, and it would be nice to make this configurable.
(https://github.com/envoyproxy/envoy/blob/master/source/common/http/rest_api_fetcher.cc#L56)

### PR DESCRIPTION
**Pull request by venilnoronha**
*Description*:
The existing `RestApiFetcher` implementation used a hard-coded request timeout of 1s which wouldn't suit all environments. This commit enables the configuration of request timeout in the `RestApiFetcher`.

Signed-off-by: Venil Noronha <veniln@vmware.com>

*Risk Level*: Low
*Testing*: Updated existing tests
*Docs Changes*: N/A
*Release Notes*: N/A
Fixes #1833 

/cc @rshriram

### CODE REVIEW

**Comment by mattklein123** [comment]
@venilnoronha LMK when this is ready for review

**Comment by venilnoronha** [comment]
@mattklein123 it'd be great if you could do a quick rundown on this. I'm still trying to understand the v1/v2 structuring, so any hints on that would be helpful.

File: source/common/config/utility.cc [hunk 1]
```diff
@@ -159,10 +161,21 @@ std::chrono::milliseconds Utility::apiConfigSourceRefreshDelay(
       DurationUtil::durationToMilliseconds(api_config_source.refresh_delay()));
 }
 
+std::chrono::milliseconds Utility::apiConfigSourceRequestTimeout(
+    const envoy::api::v2::core::ApiConfigSource& api_config_source) {
+  if (!api_config_source.has_request_timeout()) {
+    throw EnvoyException("request_timeout is required for REST API configuration sources");
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/common/config/utility.cc]: +1, please just use a default of 1s and then revert all the example config changes in the PR.

[the remark this answers is not in the document]
Comment (venilnoronha) [reply source/common/config/utility.cc]: Addressed in bb245a9.

File: source/common/json/config_schemas.cc [hunk 2]
```diff
@@ -220,6 +220,11 @@ const std::string Json::Schema::CLIENT_SSL_NETWORK_FILTER_SCHEMA(R"EOF(
         "type" : "integer",
         "minimum" : 0,
         "exclusiveMinimum" : true
+      },
+      "request_timeout_ms" : {
```
Comment (mattklein123) [inline source/common/json/config_schemas.cc]: Revert all changes to the v1 schemas. Basically, v1 should be hard coded to 1s.

File: api/envoy/config/filter/network/client_ssl_auth/v2/client_ssl_auth.proto [hunk 3]
```diff
@@ -30,6 +30,9 @@ message ClientSSLAuth {
   // 0-refresh_delay_ms milliseconds.
   google.protobuf.Duration refresh_delay = 3 [(gogoproto.stdduration) = true];
 
+  // Time in milliseconds for an authentication service request to timeout. Default is 1000 (1s).
+  google.protobuf.Duration request_timeout = 5 [(gogoproto.stdduration) = true];
```
Comment (mattklein123) [inline api/envoy/config/filter/network/client_ssl_auth/v2/client_ssl_auth.proto]: TBH I would revert all the changes for this filter until someone asks for this.

File: source/common/config/filter_json.cc [hunk 4]
```diff
@@ -433,6 +433,7 @@ void FilterJson::translateClientSslAuthFilter(
   JSON_UTIL_SET_STRING(json_config, proto_config, auth_api_cluster);
   JSON_UTIL_SET_STRING(json_config, proto_config, stat_prefix);
   JSON_UTIL_SET_DURATION(json_config, proto_config, refresh_delay);
+  JSON_UTIL_SET_DURATION(json_config, proto_config, request_timeout);
```
Comment (mattklein123) [inline source/common/config/filter_json.cc]: recommend reverting

File: source/common/config/utility.cc [hunk 5]
```diff
@@ -40,6 +40,8 @@ void Utility::translateApiConfigSource(const std::string& cluster, uint32_t refr
 
   api_config_source.mutable_refresh_delay()->CopyFrom(
       Protobuf::util::TimeUtil::MillisecondsToDuration(refresh_delay_ms));
+  api_config_source.mutable_request_timeout()->CopyFrom(
```
Comment (mattklein123) [inline source/common/config/utility.cc]: Since v1 is hard coded effectively, can we set 1s here and hard code it? Or just leave empty and then when the proto is pulled from it will just default to 1s? So I'm thinking almost all these changes can be reverted.

File: source/common/json/config_schemas.cc (hunk 2 above)
Comment (venilnoronha) [reply source/common/json/config_schemas.cc]: Addressed in 0ecfc41.

File: api/envoy/config/filter/network/client_ssl_auth/v2/client_ssl_auth.proto (hunk 3 above)
Comment (venilnoronha) [reply api/envoy/config/filter/network/client_ssl_auth/v2/client_ssl_auth.proto]: Addressed in 0ecfc41.

File: source/common/config/filter_json.cc (hunk 4 above)
Comment (venilnoronha) [reply source/common/config/filter_json.cc]: Addressed in 0ecfc41.

File: source/common/config/utility.cc (hunk 5 above)
Comment (venilnoronha) [reply source/common/config/utility.cc]: Addressed in 0ecfc41.

File: source/common/router/rds_subscription.cc [hunk 6]
```diff
@@ -20,6 +20,8 @@ RdsSubscription::RdsSubscription(
     : RestApiFetcher(cm, rds.config_source().api_config_source().cluster_names()[0], dispatcher,
```
Comment (mattklein123) [inline source/common/router/rds_subscription.cc]: Does everywhere that has a RestApiFetcher have an api_config_source at this point? Can we just pass the API config source to the fetcher? I think that would make a bunch of callsites simpler and push the other logic inside the fetcher?

File: source/common/config/utility.cc (hunk 5 above)
Comment (venilnoronha) [reply source/common/config/utility.cc]: Yes, I've fixed this in 8f47c5b.

File: api/envoy/api/v2/core/config_source.proto [hunk 7]
```diff
@@ -44,6 +44,9 @@ message ApiConfigSource {
 
   // For REST APIs, the delay between successive polls.
   google.protobuf.Duration refresh_delay = 3 [(gogoproto.stdduration) = true];
+
+  // For REST APIs, the request timeout.
```
[the remark this answers is not in the document]
Comment (venilnoronha) [reply api/envoy/api/v2/core/config_source.proto]: Addressed in 115a8e7.

File: api/envoy/api/v2/core/config_source.proto [hunk 8]
```diff
@@ -44,6 +44,9 @@ message ApiConfigSource {
...
+  google.protobuf.Duration request_timeout = 5 [(gogoproto.stdduration) = true];
```
[the remark this answers is not in the document]
Comment (venilnoronha) [reply api/envoy/api/v2/core/config_source.proto]: Addressed in 115a8e7.

File: source/common/router/rds_subscription.cc (hunk 6 above)
Comment (venilnoronha) [reply source/common/router/rds_subscription.cc]: Addressed in 1627b53.

File: source/common/config/utility.cc [hunk 9]
```diff
@@ -159,6 +159,16 @@ std::chrono::milliseconds Utility::apiConfigSourceRefreshDelay(
       DurationUtil::durationToMilliseconds(api_config_source.refresh_delay()));
 }
 
+std::chrono::milliseconds Utility::apiConfigSourceRequestTimeout(
+    const envoy::api::v2::core::ApiConfigSource& api_config_source) {
+  if (!api_config_source.has_request_timeout()) {
```
Comment (mattklein123) [inline source/common/config/utility.cc]: Sorry forgot to comment on this before but I think you can shorten this to 1 line with PROTOBUF_GET_OPTIONAL_MS(...) (search code)

File: test/config/integration/server.yaml [hunk 10]
```diff
@@ -429,11 +429,13 @@ dynamic_resources:
       cluster_names:
       - lds
       refresh_delay: 30s
+      request_timeout: 1s
```
Comment (mattklein123) [inline test/config/integration/server.yaml]: I would revert the changes in this file.

File: test/common/config/subscription_factory_test.cc [hunk 11]
```diff
@@ -91,6 +91,7 @@ TEST_F(SubscriptionFactoryTest, RestClusterSingleton) {
 
   config.mutable_api_config_source()->set_api_type(envoy::api::v2::core::ApiConfigSource::REST);
   config.mutable_api_config_source()->mutable_refresh_delay()->set_seconds(1);
+  config.mutable_api_config_source()->mutable_request_timeout()->set_seconds(1);
```
Comment (mattklein123) [inline test/common/config/subscription_factory_test.cc]: Now that we have the prof code change locked down, can you audit all the test changes? In general I don't think you need to set the default all over the place. I would just make sure that you have a test that tests something other than the default in at least one place. The rest can probably be reverted.

**Comment by venilnoronha** [comment]
@mattklein123 thanks for the detailed review! I've addressed the remaining comments in 1f8d3f7.

**Comment by mattklein123** [comment]
@venilnoronha oops right after I merged this I realized this change should have a release note. If you don't mind could you add one in a follow up PR? Thank you!

**Comment by venilnoronha** [comment]
@mattklein123 of course! Thanks for merging!
